c/dao.c

```diff
@@ -0,0 +1,290 @@
+ * of witness table. Kept as a separate function so witness buffer
+ * can be cleaned as soon as it is not needed.
+ */
+static int extract_withdraw_header_index(size_t input_index, size_t *index) {
+  int ret;
+  volatile uint64_t len = 0;
+  unsigned char witness[WITNESS_SIZE];
+
+  len = WITNESS_SIZE;
+  ret = ckb_load_witness(witness, &len, 0, input_index, CKB_SOURCE_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+  ns(Witness_table_t) witness_table = ns(Witness_as_root(witness));
+  if (witness_table == NULL) {
+    return ERROR_ENCODING;
+  }
+  ns(Bytes_vec_t) data = ns(Witness_data(witness_table));
+  if (ns(Bytes_vec_len(data)) < 2) {
+    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  }
+
+  volatile uint64_t value = 0xFFFFFFFFFFFFFFFF;
+  volatile size_t s = 8;
+  ret = extract_bytes(ns(Bytes_vec_at(data, 1)), ((unsigned char *)&value), &s);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (s != 8) {
+    return ERROR_ENCODING;
+  }
+  *index = value;
+  return CKB_SUCCESS;
+}
+
+static int calculate_dao_input_capacity(size_t input_index,
+                                        uint64_t original_capacity,
+                                        uint64_t *calculated_capacity) {
+  int ret;
+  volatile uint64_t len = 0;
+  size_t withdraw_index = 0;
+
+  ret = extract_withdraw_header_index(input_index, &withdraw_index);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  unsigned char deposit_header_buffer[HEADER_SIZE];
+  len = HEADER_SIZE;
+  ret = ckb_load_header(deposit_header_buffer, &len, 0, input_index,
+                        CKB_SOURCE_INPUT);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  unsigned char withdraw_header_buffer[HEADER_SIZE];
+  len = HEADER_SIZE;
+  ret = ckb_load_header(withdraw_header_buffer, &len, 0, withdraw_index,
+                        CKB_SOURCE_DEP);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  ns(Header_table_t) deposit_header = ns(Header_as_root(deposit_header_buffer));
+  if (deposit_header == NULL) {
+    return ERROR_ENCODING;
+  }
+  ns(Header_table_t) withdraw_header =
+      ns(Header_as_root(withdraw_header_buffer));
+  if (withdraw_header == NULL) {
+    return ERROR_ENCODING;
+  }
+
+  uint64_t deposit_number = ns(Header_number(deposit_header));
+  uint64_t withdraw_number = ns(Header_number(withdraw_header));
+
+  if (withdraw_number <= deposit_number) {
+    return ERROR_INVALID_WITHDRAW_BLOCK;
+  }
+
+  uint64_t windowleft = LOCK_PERIOD_BLOCKS -
+                        (withdraw_number - deposit_number) % LOCK_PERIOD_BLOCKS;
+  uint64_t minimal_since =
+      withdraw_number + MAX(MATURITY_BLOCKS, windowleft) + 1;
+
+  volatile uint64_t input_since = 0;
+  len = 8;
+  ret = ckb_load_input_by_field(((unsigned char *)&input_since), &len, 0, input_index,
+                                CKB_SOURCE_INPUT, CKB_INPUT_FIELD_SINCE);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+
+  if (input_since < minimal_since) {
```

Comment:
🐞This does not check the `input_slice` flags byte https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0017-tx-valid-since/0017-tx-valid-since.md